src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -517,7 +500,31 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
                 }
             }
 
-            LogMessages.OpenedPhysicalConnection(ConnectionLogger, Host, Port, Database, UserFacingConnectionString, stopwatch.ElapsedMilliseconds, Id);
+            if (_dataSource.AsyncConnectionInitializer is not null)
+            {
+                Debug.Assert(_dataSource.SyncConnectionInitializer is not null);
+
+                var tempConnection = new NpgsqlConnection(_dataSource, this);
+
+                try
+                {
+                    if (async)
+                        await _dataSource.AsyncConnectionInitializer(tempConnection);
+                    else if (!async)
+                        _dataSource.SyncConnectionInitializer(tempConnection);
+                }
+                finally
+                {
+                    // Note that we can't just close/dispose the NpgsqlConnection, since that puts the connector back in the pool.
```

Comment:
AFAIK if a connection is broken while in the middle of `AsyncConnectionInitializer`, that's actually going to happen...

Reply:
Right, but is that a problem? The exception will bubble out of the connection initializer and out of Open, and the broken connector will be "returned" to the pool, which will discard it, no?

Reply:
OK, I modified the tests to exercise this scenario.

You're right that we go into Break twice - but the 2nd Break is a no-op because the state is already Broken from the 1st.

However... It's true that in this case the connector is no longer bound, and so Connection may be null in the finally (fixed).

More importantly, our assertion that CloseConnector must find the connector in the Connectors array fails, because it hasn't been added yet (this only happens after successful opening of the connector in PoolingDataSource.OpenNewConnector. I've removed this assertion - see the comment I've added.

How does this all look?

BTW I'm no longer sure why we have to return the connector to the pool at the moment from Break... It seems like it could be cleaner/more consistent if we only returned it when NpgsqlConnection is closed (as usual). That would simplify these nested scenarios and allow us to keep the above assertion to catch bugs.

Reply:
> BTW I'm no longer sure why we have to return the connector to the pool at the moment from Break... It seems like it could be cleaner/more consistent if we only returned it when NpgsqlConnection is closed (as usual). That would simplify these nested scenarios and allow us to keep the above assertion to catch bugs.

https://github.com/npgsql/npgsql/issues/3705#issuecomment-839908772 TLDR: efcore :bee:

Reply:
Great, good thing one of us has the history of the project in their head :)

We could instead consider just leaving the connection state as Open, rather than Closed as discussed in that issue; that's even maybe correct (in the sense that it still needs to be closed). But it could again cause issues in upper layers which expect an open connection to be functional...

Anyway, I'll add a comment explaining this so i don't forget again ;)